Fix for Issue #253 - Features: Disassembler, Cross-platform Disassembler, Cross-profiling settings dialog

dknysh opened the pull request:
Hello Milian,

Could you please review and merge the changes for Issue #253?
Following our discussion about the features review and merge process I have splitted the all changes on 21 parts.

Thank you,
Darya

milianw commented:
Just a note: I've reviewed the perfparser change for now, which should be enough of work for you for starters. Once that's done, I'll also have a look at this here.

Thanks!

dknysh commented:
Could you give me any feedback please?

milianw commented on src/models/filterandzoomstack.cpp:
```diff
@@ -34,6 +34,8 @@
 FilterAndZoomStack::FilterAndZoomStack(QObject* parent)
     : QObject(parent)
 {
+    m_actions.disassembly = new QAction(QIcon::fromTheme(QStringLiteral("view-filter")), tr("Disassembly"), this);
```
showing the disassembly is not a filter operation

also, as icon, what about using view-list-details instead?

milianw commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -74,18 +74,35 @@ struct StringId
     qint32 id = -1;
 };
 
+struct StringId64
```
I don't understand why this is needed - is the string id suddenly 64bit? if so, then the non-64 string id needs to be removed _everywhere_

milianw replied:
but, more probably, this indicates a bug in perfparser that now writes 64bit string ids somewhere? these should remain at 32bit

milianw replied:
ah, looking down I realize where this comes from, remove this class please and use `quint64` directly for the addresses below

milianw commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -215,20 +232,24 @@ QDebug operator<<(QDebug stream, const LocationDefinition& locationDefinition)
 struct Symbol
 {
     StringId name;
+    StringId64 relAddr;
```
huh, these are both not string ids, these are just addresses, both of these members should be plain quint64

milianw commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -921,9 +942,11 @@ class PerfParserPrivate : public QObject
         Q_ASSERT(bottomUpResult.symbols.size() > symbol.id);
         // TODO: isKernel information
         const auto symbolString = strings.value(symbol.symbol.name.id);
+        const auto relAddrString = symbol.symbol.relAddr.id;
```
these should not be strings, these should be the address/size values directly

milianw commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -1144,6 +1167,7 @@ class PerfParserPrivate : public QObject
         summaryResult.totalMemoryInKiB = features.totalMem;
 
         eventResult.cpus.resize(features.nrCpusAvailable);
+        disassemblyResult.arch = QString::fromUtf8(features.arch);
```
store this in a separate member `m_arch`

dknysh replied:
Do you mean rename the field on m_arch instead arch?

milianw replied:
no, I mean remove the `disassemblyResult` as a member and instead introduce a separate `m_arch` member, then see below...

milianw replied:
this is still here, please clean it up

dknysh replied:
It is quite small to be hoisted into separate patch. I just added it to the actualPath patch because the both are related to some extent with cross-platformity.

milianw replied:
please split it up, commits should be atomic and not do multiple things that are only slightly related

milianw commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -1303,6 +1328,8 @@ void PerfParser::startParseFile(const QString& path, const QString& sysroot, con
     m_bottomUpResults = {};
     m_callerCalleeResults = {};
     m_events = {};
+    m_disassemblyResult = {};
```
```
m_disassemblyResult = {m_arch, path, appPath, extraLibPaths};
if (!arch.isEmpty())
    m_disassemblyResult.arch = arch;
```

milianw replied:
see here how to use `m_arch` then as an example

milianw commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -1342,6 +1369,10 @@ void PerfParser::startParseFile(const QString& path, const QString& sysroot, con
                         emit topDownDataAvailable(d.topDownResult);
                         emit summaryDataAvailable(d.summaryResult);
+
+                        d.disassemblyResult.copy(m_disassemblyResult);
```
why is there both `d.disassemblyResult` and `m_disassemblyResult`?

milianw replied:
and this, afaik if you just keep the arch you can remove this one and all the strange functions on the disassembly result struct itself

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+#include "resultsdisassemblypage.h"
```
missing license header

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -0,0 +1,53 @@
+#pragma once
```
missing license header

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+
+ResultsDisassemblyPage::~ResultsDisassemblyPage() = default;
+
+/**
```
useless comment, remove

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+    }
+}
+
+/**
```
move the documentation to the header please

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+/**
+ *  Produce and show disassembly with 'objdump'
+ */
+void ResultsDisassemblyPage::showDisassembly() {
```
useless function as it just forwards to byAddressRange ? why not unify the two?

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+ *  Produce disassembler with 'objdump' and output to Disassembly tab
+ */
+void ResultsDisassemblyPage::showDisassembly(QString processName) {
+    QTemporaryFile m_tmpFile;
```
ideally all of this code should be rewritten to be async using signals/slots, instead of blocking the UI while we are waiting for objdump to finish

milianw replied:
i.e. show a busy indicator while we are waiting for the data to come in, but otherwise don't block the UI

milianw replied:
furthermore, please move the logic to run and parse objdump into a separate class that can be tested separately from the UI

dknysh replied:
> i.e. show a busy indicator while we are waiting for the data to come in, but otherwise don't block the UI

Can this be refactored in this async way after all patches are merged. It is extremely complicated for me to sync with the changes in other commits when I change in a particular commit. Do you insist such async refactoring or it can be postponed?

milianw replied:
ok, that can be postponed, but it is very important to have

milianw replied:
please add at least a TODO that this needs to be extracted into a standalone function, covered by unit tests and made async

dknysh replied:
Ok, I will add

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    // Call objdump with arguments: addresses range and binary file
+    QString processName =
```
don't merge the process name and its args, use the QProcess api that takes (QString, QStringList) separately

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+/**
+ *  Produce disassembler with 'objdump' and output to Disassembly tab
+ */
+void ResultsDisassemblyPage::showDisassembly(QString processName) {
```
const&

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+    m_extraLibPaths = data.extraLibPaths;
+    m_arch = data.arch.trimmed().toLower();
+
+    m_objdump = m_arch.startsWith(QLatin1String("arm")) ? QLatin1String("arm-linux-gnueabi-objdump") : QLatin1String(
```
we need validation via `QStandardPaths::findExecutable` if either of these is found. if not, this whole feature needs to be disabled with an error message saying that the process isn't found

also: QStringLiteral instead of QLatin1String here

milianw replied:
furthermore, we need to give the user the ability to configure the arch <-> objdump mapping somehow in the settings

dknysh replied:
> we need validation via `QStandardPaths::findExecutable` if either of these is found. if not, this whole feature needs to be disabled with an error message saying that the process isn't found

How should I disable the whole feature when it was not found? 
Is it not enough instead to check if QProcess was started successfully and advise to install appropriate objdump if was not?

milianw replied:
ok yes that could work too, but make sure that the error message is descriptive enough in such a case

milianw replied:
oh and don't forget to add objdump as a runtime dependency in cmake, i.e. `find_package` + `set_package_properties` with `TYPE RUNTIME`

milianw replied:
can you please at least add a TODO for the configuration of the objdump path/name?

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+    }
+}
+
+/**
```
wrong/useless comment?

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+    }
+}
+
+/**
```
rename to setSymbol and remove the useless comment, or expand it

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+
+    if (m_tmpFile.open()) {
+        int row = 0;
+        QStandardItemModel *model = new QStandardItemModel();
```
please don't use QStandardItemModel, implement a real one (in a separate file) that operates on a vector of structs that hold the actual data

dknysh replied:
I enter DisassemblyModel in commit e68b7b7d0c5515c4b5e69f8058d1b31fa0c5ac2d. Please consider QStandardItemModel  as temporary.

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,159 @@
+    if (m_curSymbol.symbol.isEmpty()) {
+        return;
+    }
+    // If binary is not found at the specified path, use current binary file located at the application path
```
all of this mapping logic looks like it should be done in perfparser instead, it knows where a file resides after all (that's where it parses the debug info from e.g.)

dknysh replied:
How do you suggest to pass it to hotspot from perfparser? Enter additional Data::Symbol field for actual path?  Probably it is better to compute it here in hotspot? Do you insist I move the mapping into perfparser?

milianw replied:
yes, ideally we add another path to the symbol with the path to the mapped file

and yes, I insist on that, as it would be required for anyone else consuming the new perfparser feature to integrate a dissasembly viewer

milianw reviewed (changes requested):
sorry, I'm quite busy - but I took the time to review the first patch at least now.

once you split out the running+parsing of objdump, please add a test for running that on one of the symbols of the precompiled binaries that are included in the perfparser test directory please?

dknysh commented on src/mainwindow.cpp:
```diff
@@ -182,7 +185,11 @@ MainWindow::MainWindow(QWidget* parent)
+
+    QList<QAction *> actions = m_resultsPage->filterMenu()->actions();
+    // Disassembly menu items should not be presented here
+    actions.removeFirst();
```
I thought it is rather natural to add it into context menu. Do you see better alternative?

milianw replied:
oh, certainly it should be part of the context menu, but as a separate action - not as part of the *filter menu*. this may require some other changes to make this new action fit in more naturally. but it has nothing to do with filtering. it's more like the other action to switch to the caller/callee view, have a look at how that is done please

dknysh replied:
Ok.

dknysh replied:
Don't you mind I keep Action disassembly in FilterAndZoomStack? It is used everywhere in setupContextMenu. For all pages.

milianw replied:
I don't see/understand why this would be necessary - just follow the approach taken by `jumpToCallerCallee` and be done with it no?

dknysh replied:
To be able to call Disassembly from any tab - Bottom Up,  Top down, Caller / Callee through context menu of selected symbol. There is the method setupContextMenu of ResultUtil which is called for each tab. It has argument FilterAndZoomStack* filterStack. It seems to me better to add Disassembly action to the structure than propagate separately this action for every constructor for every class. I didn't catch your idea how to avoid this.

dknysh commented:
Updated according to your remarks. Could you please review?
As for that part about actual paths that should be proceeded in perfparser - please look at the last commit of https://github.com/KDAB/perfparser/pull/14, I have committed there the changes.
As for movement of run and parse objdump functionality into separate class - 
I am going to do this later with last commit as refactoring. It would better for me to finish review of other commits and then as refactoring move approved code into separate class.

milianw commented on CMakeLists.txt:
```diff
@@ -52,6 +52,7 @@ find_package(Elfutils REQUIRED)
 find_package(ECM 1.0.0 NO_MODULE REQUIRED)
 find_package(Threads REQUIRED)
 find_package(Zstd)
+find_package(objdump)
```
the Findobjdump.cmake is missing from this patch

also, you need to add the RUNTIME dependency somewhere too, no? here's an example from a different project:

```

find_package(heaptrack QUIET)
set_package_properties(heaptrack PROPERTIES
    DESCRIPTION "A heap memory profiler for Linux"
    URL "https://phabricator.kde.org/dashboard/view/28"
    PURPOSE "Recommended: required by the non-essential heaptrack plugin"
    TYPE RUNTIME
)
```

something like that for objdump will be needed here too

milianw commented on src/models/data.h:
```diff
@@ -58,6 +60,10 @@ struct Symbol
     // prettified function name
     QString prettySymbol;
+    // relative address
+    quint64 relAddr;
```
= 0, also below

dknysh replied:
Sorry, what do you mean? I set default value 0 for it.

milianw commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -215,20 +215,24 @@ QDebug operator<<(QDebug stream, const LocationDefinition& locationDefinition)
 struct Symbol
 {
     StringId name;
+    quint64 relAddr;
```
both = 0

dknysh replied:
Sorry, missed the remark, I will add into next iteration of review

milianw commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -1303,6 +1311,9 @@ void PerfParser::startParseFile(const QString& path, const QString& sysroot, con
     m_bottomUpResults = {};
     m_callerCalleeResults = {};
     m_events = {};
+    m_disassemblyResult = {path, appPath, extraLibPaths, arch};
```
m_arch here

milianw replied:
or even better:

`arch.isEmpty() ? m_arch : arch`

dknysh replied:
There is no m_arch in this class. It is within PerfParser class method. But m_arch is field of PerfParserPrivate. It is computed in void setFeatures(const FeaturesDefinition& features) method of PerfParserPrivate

milianw replied:
ah, now I see - the arch here is the arg passed in. then just remove the two new lines below, they make no sense. you are setting the arch already, so setting it again when not empty (to itself) makes no sense?!

dknysh replied:
Ok, I will at next iteration

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,144 @@
+  This file is part of Hotspot, the Qt GUI for performance analysis.
+
+  Copyright (C) 2017-2020 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+  Author: Nate Rogers <nate.rogers@kdab.com>
```
:) your name here please

dknysh replied:
Ok)

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,144 @@
+#include <QStandardItemModel>
+
+ResultsDisassemblyPage::ResultsDisassemblyPage(FilterAndZoomStack *filterStack, PerfParser *parser, QWidget *parent)
+        : QWidget(parent), ui(new Ui::ResultsDisassemblyPage) {
```
is this what you got from clang-format? I don't think so, or? it should be:

````
Foo::Foo(...)
    : a(...)
    , b(...)
{
    ...
}
```

dknysh replied:
Hm, I have run git-clang-format HEAD^ before commit and then commit as you asked. Ok, I will fix this manually.

milianw replied:
if this is clang-format clean, then leave it as is, I'll have to fix the clang format style file then

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,144 @@
+
+ResultsDisassemblyPage::~ResultsDisassemblyPage() = default;
+
+void ResultsDisassemblyPage::clear() {
```
the brace is on the wrong line here too

milianw replied:
repeats below

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,144 @@
+ResultsDisassemblyPage::~ResultsDisassemblyPage() = default;
+
+void ResultsDisassemblyPage::clear() {
+    if (ui->asmView->model() == nullptr) return;
```
return on next line

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -0,0 +1,79 @@
+  This file is part of Hotspot, the Qt GUI for performance analysis.
+
+  Copyright (C) 2017-2020 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+  Author: Nate Rogers <nate.rogers@kdab.com>
```
:) also wrong

dknysh replied:
I have added TODO for 'needs to be extracted into a standalone function, covered by unit tests and made async' to commit e2cf15317f13db10516472d8b08e04f9b0f5644f

milianw reviewed (changes requested):
here some remarks for the first commit in this series, will try to find the time for the others

next

milianw commented on src/qwebstyleedit.h:
```diff
@@ -0,0 +1,39 @@
+    void setText(const QString&);
+    QString text() const;
+    void setGrayedText(const QString&);
+    QString grayedText() const;
```
no get prefixes in getters

milianw commented on src/settingsdialog.h:
```diff
@@ -0,0 +1,43 @@
+#ifndef SETTINGSDIALOG_H
```
missing license, use pragma once

milianw commented on src/settingsdialog.h:
```diff
@@ -0,0 +1,43 @@
+#include <QDialog>
+#include <QDialogButtonBox>
+#include <QAbstractButton>
+#include "qwebstyleedit.h"
```
newline before

milianw commented on src/settingsdialog.h:
```diff
@@ -0,0 +1,43 @@
+    Q_OBJECT
+
+public:
+    MainWindow* mainWindow;
```
make private, = nullptr

milianw replied:
remove, the dialog should be fed from the mainwindow with the parameters and then we query back for the new values when the dialog got accepted (e..g via a signal)

don't set the params directly on the mainwindow from the settings dialog

milianw commented on src/settingsdialog.h:
```diff
@@ -0,0 +1,43 @@
+    MainWindow* mainWindow;
+    explicit SettingsDialog(QWidget* parent);
+    ~SettingsDialog();
+    void setTexts(QWebStyleEdit, const QString&, const QString&, const QString&);
```
missing names for the arguments

milianw commented on src/startpage.cpp:
```diff
@@ -118,3 +117,7 @@ void StartPage::updateBackground()
+
+void StartPage::on_pathsAndArchSettingsButton_clicked() {
+    emit pathsAndArchSettingsButtonClicked();
+}
```
this isn't connected anywhere, is it? generally I'd prefer if you remove this and just connect the signal of the button to the signal of the StartPage directly

dknysh commented:
Updated. Moved out the changes concerned Cross-profiling settings dialog.  
Please review.

milianw commented on src/resultscallercalleepage.h:
```diff
@@ -65,6 +65,7 @@ private slots:
 signals:
     void navigateToCode(const QString& url, int lineNumber, int columnNumber);
     void navigateToCodeFailed(const QString& message);
+    void jumpToDisassembly(const Data::Symbol &symbol);
```
& should be next to the type

milianw commented on src/resultspage.cpp:
```diff
@@ -90,6 +92,11 @@ ResultsPage::ResultsPage(PerfParser* parser, QWidget* parent)
+    ui->resultsTabWidget->addTab(m_resultsDisassemblyPage, tr("Disassembly"));
+
+    int tabsCount = ui->resultsTabWidget->count();
+    ui->resultsTabWidget->removeTab(tabsCount - 1);
```
why are you removing the tab here again?

milianw replied:
just don't add it in the first place, no?

milianw commented on src/resultspage.cpp:
```diff
@@ -212,6 +230,24 @@ void ResultsPage::onJumpToCallerCallee(const Data::Symbol& symbol)
+
+void ResultsPage::onJumpToDisassembly(const Data::Symbol& symbol)
+{
+    ui->resultsTabWidget->addTab(m_resultsDisassemblyPage, tr("Disassembly"));
```
what happens if this is called twice, for different symbols? ideally we should be able to show both to compare them

but for now just one would be fine, just make sure that we don't end up with two that show the same data?

milianw replied:
this is still not answered?

dknysh replied:
Evething is OK for this. Each of them will show its own Disassemby data.

milianw replied:
how is that possible?  there is only one m_resultsDisassemblyPage, and it gets added multiple times? So I would assume that there will be multiple tabs all showing the same data?

for this to work as you explain, m_resultsDisassemblyPage would need to become a vector, and new pages added here?

milianw replied:
this has not been answered

milianw commented on src/resultspage.cpp:
```diff
@@ -212,6 +230,24 @@ void ResultsPage::onJumpToCallerCallee(const Data::Symbol& symbol)
     ui->resultsTabWidget->setCurrentWidget(m_resultsCallerCalleePage);
 }
 
+void ResultsPage::setData(const Data::Symbol& data)
```
the name of this function is pretty bad, I would just remove it and inline the call below instead to where it is currently being called -  that it makes it clear what this is doing

milianw commented on src/resultspage.cpp:
```diff
@@ -212,6 +230,24 @@ void ResultsPage::onJumpToCallerCallee(const Data::Symbol& symbol)
+    m_resultsDisassemblyPage->setSymbol(data);
+}
+
+void ResultsPage::setData(const Data::DisassemblyResult& data)
```
I also think that this shouldn't be a public function, it's only called internally, no? just inline the call below

milianw commented on src/resultspage.cpp:
```diff
@@ -126,6 +133,11 @@ ResultsPage::ResultsPage(PerfParser* parser, QWidget* parent)
     });
+
+    connect(parser, &PerfParser::disassemblyDataAvailable, this, [this](const Data::DisassemblyResult& data) {
+        this->setData(data);
```
see below, this should instead be:

```
connect(parser, &PerfParser::disassemblyDataAvailable, m_resultsDisassemblypage, &...::setData);
```
```

milianw commented on src/resultspage.cpp:
```diff
@@ -212,6 +230,24 @@ void ResultsPage::onJumpToCallerCallee(const Data::Symbol& symbol)
+{
+    ui->resultsTabWidget->addTab(m_resultsDisassemblyPage, tr("Disassembly"));
+    ui->resultsTabWidget->setCurrentWidget(m_resultsDisassemblyPage);
+    setData(symbol);
```
see above, just call

```
m_resultsDisassemblypage->setSymbol(symbol);
```

here instead

milianw commented on src/resultspage.h:
```diff
@@ -28,6 +28,7 @@
 #pragma once
 
 #include <QWidget>
+#include "data.h"
```
remove, use forward declarations instead where needed

milianw commented on src/resultsutil.h:
```diff
@@ -74,6 +74,7 @@ enum class CallbackAction
 {
     ViewCallerCallee = 0x1,
     OpenEditor = 0x2,
+    ViewDisassembly = 0x3
```
= 0x4

otherwise `ViewDisassembly` is equal to `ViewCallerCallee | OpenEditor`

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -115,19 +124,52 @@ void ResultsDisassemblyPage::showDisassembly(QString processName, QStringList ar
 
+            QStringList asmTokens = asmLine.split(QLatin1Char(':'));
+            QString addrLine = asmTokens.value(0);
+            QString costLine = QString();
```
looking below, this line should be removed

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -115,19 +124,52 @@ void ResultsDisassemblyPage::showDisassembly(QString processName, QStringList ar
+
+            // Calculate event times and add them in red to corresponding columns of the current disassembly row
+            for (int event = 0; event < m_disasmResult.selfCosts.numTypes(); event++) {
+                float totalCost = 0;
```
float costLine = 0;

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -115,19 +124,52 @@ void ResultsDisassemblyPage::showDisassembly(QString processName, QStringList ar
+                    Data::LocationCost locationCost = i.value();
+                    float cost = locationCost.selfCost[event];
+                    if (QString::number(location.relAddr, 16) == addrLine.trimmed()) {
+                        costLine = QString::number(cost);
```
costLine = cost

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -115,19 +124,52 @@ void ResultsDisassemblyPage::showDisassembly(QString processName, QStringList ar
+                    i++;
+                }
+
+                float costInstruction = costLine.toFloat();
```
not needed when you keep it a float to begin with

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -115,19 +124,52 @@ void ResultsDisassemblyPage::showDisassembly(QString processName, QStringList ar
+                    ? QString::number(costInstruction * 100 / totalCost, 'f', 2) + QLatin1String("%")
+                    : QString();
+
+                QStandardItem* costItem = new QStandardItem(costLine);
```
you can setData(Qt::DisplayRole, costLine) (i.e. float -> QVariant), no need to convert to string here

milianw replied:
oh wait, you are trying to replicate what the real models are doing with qstandarditemmodel... sigh - as I said this really shouldn't be used and I hope you will clean this up in the future

anyhow, please still simplify the cost -> string -> float -> string conversion and only keep the final step

milianw replied:
and add a big fat FIXME that the QStandardItem stuff must be reimplemented properly

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -27,9 +27,10 @@
-#include <QWidget>
-#include <QTemporaryFile>
 #include "data.h"
+#include <QStandardItemModel>
```
forward declare

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -27,9 +27,10 @@
-#include <QTemporaryFile>
 #include "data.h"
+#include <QStandardItemModel>
+#include <QTemporaryFile>
```
not needed in header

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -27,9 +27,10 @@
 #include "data.h"
+#include <QStandardItemModel>
+#include <QTemporaryFile>
+#include <QWidget>
```
also remove

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,153 @@
+
+    if (m_tmpFile.open()) {
+        int row = 0;
+        QStandardItemModel *model = new QStandardItemModel();
```
you are leaking the model, pass a parent:

auto *model = new QStandardItemModel(this);

better yet make it a member and clear it instead of creating a new model all the time for every time this function is called

dknysh replied:
I have fixed this in later commit 16ac4aa6089c4c4b165f488ceff9b063a5880ba3

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+    m_callStack.clear();
+}
+
+void calcFunction(QString asmLineCall, QString* offset, QString* symName)
```
static or move into anon namespace

milianw replied:
also, return the strings in a struct instead of using out parameters

milianw replied:
and take the asmLineCall by const&

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+
+void calcFunction(QString asmLineCall, QString* offset, QString* symName)
+{
+    QStringList sym = asmLineCall.trimmed().split(QLatin1Char('<'));
```
please document an expected string line and the expected parsed output

milianw replied:
remove the trimmed() (not needed as you can trim the results below anyways), use splitRef

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+void calcFunction(QString asmLineCall, QString* offset, QString* symName)
+{
+    QStringList sym = asmLineCall.trimmed().split(QLatin1Char('<'));
+    *offset = sym[0].trimmed();
```
will crash if the line is empty, use .value()

dknysh replied:
I have fixed this in one of next commits.

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+{
+    QStringList sym = asmLineCall.trimmed().split(QLatin1Char('<'));
+    *offset = sym[0].trimmed();
+    *symName = sym[1].replace(QLatin1Char('>'), QLatin1String(""));
```
dito, use value().trimmed() and then .remove() instead of replacing with an empty string

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+    *symName = sym[1].replace(QLatin1Char('>'), QLatin1String(""));
+}
+
+void ResultsDisassemblyPage::jumpToAsmCallee(QModelIndex index)
```
const&

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+
+void ResultsDisassemblyPage::jumpToAsmCallee(QModelIndex index)
+{
+    QStandardItem* asmItem = model->takeItem(index.row(), 0);
```
why are you taking the item?

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+void ResultsDisassemblyPage::jumpToAsmCallee(QModelIndex index)
+{
+    QStandardItem* asmItem = model->takeItem(index.row(), 0);
+    QString opCodeCall = m_arch.startsWith(QLatin1String("arm")) ? QLatin1String("bl") : QLatin1String("callq");
```
please parse the arch once instead of doing that repeatedly here, also use auto on the left  hand side (i.e. make the result a QLatin1String)

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+{
+    QStandardItem* asmItem = model->takeItem(index.row(), 0);
+    QString opCodeCall = m_arch.startsWith(QLatin1String("arm")) ? QLatin1String("bl") : QLatin1String("callq");
+    QString opCodeReturn = m_arch.startsWith(QLatin1String("arm")) ? QLatin1String("ret") : QLatin1String("retq");
```
dito

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+    QStandardItem* asmItem = model->takeItem(index.row(), 0);
+    QString opCodeCall = m_arch.startsWith(QLatin1String("arm")) ? QLatin1String("bl") : QLatin1String("callq");
+    QString opCodeReturn = m_arch.startsWith(QLatin1String("arm")) ? QLatin1String("ret") : QLatin1String("retq");
+    QString asmLine = asmItem->text();
```
const auto

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+
+        calcFunction(asmLineSimple.section(opCodeCall, 1), &offset, &symName);
+
+        QHash<Data::Symbol, Data::DisassemblyEntry>::iterator i = m_disasmResult.entries.begin();
```
auto

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+    QString asmLine = asmItem->text();
+    if (asmLine.contains(opCodeCall)) {
+        QString offset, symName;
+        QString asmLineSimple = asmLine.simplified();
```
inline below:

```
const auto info = calcFunction(asmLine.simplified().section(opCodeCall, 1));
// use info.offset, info.symName
```

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+
+        QHash<Data::Symbol, Data::DisassemblyEntry>::iterator i = m_disasmResult.entries.begin();
+        while (i != m_disasmResult.entries.end()) {
+            QString relAddr = QString::number(i.key().relAddr, 16);
```
instead of converting every address to a string, just convert the offset to a number in calcFunction once

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,3 +201,46 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
+            i++;
+        }
+    }
+    if (asmLine.contains(opCodeReturn)) {
```
else if (!m_callStack.isEmpty() && asmLine.contains(opCodeReturn))

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -63,9 +64,19 @@ class ResultsDisassemblyPage : public QWidget
 private:
     QScopedPointer<Ui::ResultsDisassemblyPage> ui;
+    // Asm view model
+    QStandardItemModel* model;
```
= nullptr;

milianw replied:
add m_ prefix

milianw commented on src/mainwindow.cpp:
```diff
@@ -404,11 +388,59 @@ void MainWindow::setupPathSettingsMenu()
 
-void MainWindow::setupCodeNavigationMenu()
-{
+void MainWindow::setupFilterDisassemblyMenu() {
```
{ on its own line

milianw commented on src/mainwindow.cpp:
```diff
@@ -404,11 +388,59 @@ void MainWindow::setupPathSettingsMenu()
-void MainWindow::setupCodeNavigationMenu()
-{
+void MainWindow::setupFilterDisassemblyMenu() {
+    QMenu *filterDisassemblyMenu = new QMenu(tr("Filter Disassembly View"));
```
auto*

milianw commented on src/mainwindow.cpp:
```diff
@@ -404,11 +388,59 @@ void MainWindow::setupPathSettingsMenu()
+    ui->viewMenu->addMenu(filterDisassemblyMenu);
+}
+
+void MainWindow::setupViewMenu() {
```
dito

milianw commented on src/mainwindow.cpp:
```diff
@@ -404,11 +388,59 @@ void MainWindow::setupPathSettingsMenu()
+
+    ui->viewMenu->addSeparator();
+
+    QList<QAction *> actions = m_resultsPage->filterMenu()->actions();
```
auto

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -106,6 +127,12 @@ void ResultsDisassemblyPage::showDisassembly()
 void ResultsDisassemblyPage::showDisassembly(QString processName, QStringList arguments)
 {
+    if (m_noShowRawInsn)
+        arguments << QLatin1String("--no-show-raw-insn");
```
QStringLiteral

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -106,6 +127,12 @@ void ResultsDisassemblyPage::showDisassembly()
+        arguments << QLatin1String("--no-show-raw-insn");
+
+    if (m_intelSyntaxDisassembly)
+        arguments <<  QLatin1String("-M intel ");
```
QStringLiteral

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -140,6 +167,13 @@ void ResultsDisassemblyPage::showDisassembly(QString processName, QStringList ar
             QString costLine = QString();
 
+            if (m_noShowAddress) {
+                QRegExp hexMatcher(QLatin1String("[0-9A-F]+$"), Qt::CaseInsensitive);
```
don't use QRegExp in new code, use QRegularExpression instead

pass pattern as QStringLiteral

```
static const auto hexMatcher = QRegularExpression(QStringLiteral("[0-9a-fA-F]+$"));
```

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -244,3 +278,18 @@ void ResultsDisassemblyPage::jumpToAsmCallee(QModelIndex index)
     showDisassembly();
 }
+
+void ResultsDisassemblyPage::setNoShowRawInsn(bool noShowRawInsn)
```
remove abbrevations, rename to:

```
setHideRawInstructions(bool hide)
{
    m_hideRawInstructions = hide;
    showDisassembly();
}
```

milianw replied:
this repeats below

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -244,3 +278,18 @@ void ResultsDisassemblyPage::jumpToAsmCallee(QModelIndex index)
+    m_noShowRawInsn = noShowRawInsn;
+}
+
+void ResultsDisassemblyPage::setNoShowAddress(bool noShowAddress)
```
dito: setHideAddress

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -244,3 +278,18 @@ void ResultsDisassemblyPage::jumpToAsmCallee(QModelIndex index)
+    m_noShowAddress = noShowAddress;
+}
+
+void ResultsDisassemblyPage::setIntelSyntaxDisassembly(bool intelSyntax)
```
setDisassemblySyntax(DisassemblySyntax syntax)

make the DisassemblySyntax an enum with Intel and ATT enumerators

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -91,4 +94,16 @@ public slots:
     // Map of symbols and its locations with costs
     Data::DisassemblyResult m_disasmResult;
+    // Not to show machine codes of Disassembly
+    bool m_noShowRawInsn;
```
initialize, also below

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -91,4 +94,16 @@ public slots:
+    bool m_noShowAddress;
+    // Switch Assembly to Intel Syntax
+    bool m_intelSyntaxDisassembly;
+    // Setter for m_noShowRawInsn
```
remove useless comments, move functions above the member variables to the other functions

milianw commented on src/resultspage.cpp:
```diff
@@ -270,6 +270,19 @@ void ResultsPage::clear()
     m_filterAndZoomStack->clear();
 }
 
+void ResultsPage::filterDisassemblyBytes(bool filtered)
```
use the same names as in the disassembly results page

milianw commented on src/resultspage.cpp:
```diff
@@ -270,6 +270,19 @@ void ResultsPage::clear()
+    m_resultsDisassemblyPage->filterDisassemblyBytes(filtered);
+}
+
+void ResultsPage::filterDisassemblyAddress(bool filtered) {
```
brace on its own line

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -57,7 +57,10 @@ class ResultsDisassemblyPage : public QWidget
 
     void clear();
-    void setAsmViewModel(QStandardItemModel* model, int numTypes);
+    void filterDisassemblyBytes(bool filtered);
```
remove, make the setters below public

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -63,6 +66,24 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(FilterAndZoomStack *filterStack,
 
 ResultsDisassemblyPage::~ResultsDisassemblyPage() = default;
 
+void ResultsDisassemblyPage::filterDisassemblyBytes(bool filtered)
```
remove these three functions (see below)

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -55,6 +55,9 @@
 ResultsDisassemblyPage::ResultsDisassemblyPage(FilterAndZoomStack *filterStack, PerfParser *parser, QWidget *parent)
         : QWidget(parent)
         , ui(new Ui::ResultsDisassemblyPage)
+        , m_noShowRawInsn(true)
```
move init to header

milianw commented on src/resultspage.cpp:
```diff
@@ -270,6 +270,19 @@ void ResultsPage::clear()
+    m_resultsDisassemblyPage->filterDisassemblyAddress(filtered);
+}
+
+void ResultsPage::switchOnIntelSyntax(bool intelSyntax) {
```
dito

milianw commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -1369,6 +1367,7 @@ void PerfParser::startParseFile(const QString& path, const QString& sysroot, con
             }
         });
 
+        d.disassemblyResult = m_disassemblyResult;
```
why this move?

dknysh replied:
I can not remove  disassemblyResult.arch = QString::fromUtf8(features.arch); . At least when architecture was not set through UI it should be computed automatically.

milianw replied:
I don't get that at all - the parser should know about the arch that was referenced in the `perf.data` file itself and then just pass that through. Then in the UI we can use that data to map it to a `objdump` binary, why would we need to pass an arch *into* the parser?!

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -45,12 +45,15 @@
 #include "models/treemodel.h"
 
 #include <QStandardItemModel>
+#include <QWheelEvent>
```
please remove this feature, it should be handled generically - ideally through reusing KTextEditor for showing the assembly data

milianw replied:
this would also get us proper highlighting etc. pp.

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -112,19 +137,57 @@ void ResultsDisassemblyPage::showDisassembly()
     showDisassembly(processName, arguments);
 }
 
+QByteArray ResultsDisassemblyPage::processDisassemblyGenRun(QString processName, QStringList arguments)
```
you are again doing multiple things in one commit - but as I said, remove the wheel zoom feature and then only keep the diagnostics here

milianw replied:
also: const& the function args

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -112,19 +137,57 @@ void ResultsDisassemblyPage::showDisassembly()
+{
+    QByteArray processOutput = QByteArray();
+    if (m_curSymbol.symbol.isEmpty()) {
+        processOutput = "Empty symbol ?? is selected";
```
needs to be translated, repeats below - make the error a QString instead of a QBA and split it out from the QBA of the objdump run

```
struct DisassemblyOutput
{
    QByteArray output;
    QString errorMessage;
    explicit operator bool() const { return errorMessage.isEmpty(); }
    static DisassemblyOutput fromProcess(const QString &processName, const QStringList &arguments);
};
```

milianw replied:
use a KMessageBox to display the error message then

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -60,6 +60,7 @@ class ResultsDisassemblyPage : public QWidget
     void filterDisassemblyBytes(bool filtered);
     void filterDisassemblyAddress(bool filtered);
     void switchOnIntelSyntax(bool intelSyntax);
+    QByteArray processDisassemblyGenRun(QString processName, QStringList arguments);
```
should be a free function in an anon namespace in the .cpp file

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -112,19 +137,57 @@ void ResultsDisassemblyPage::showDisassembly()
+        QProcess asmProcess;
+        asmProcess.start(processName, arguments);
+
+        bool started = asmProcess.waitForStarted();
```
simplify the huge if/else chain below via early return statements:

```
if (!asmProcess.waitForStarted()) {
    return {{}, ResultsDisassemblyPage::tr("...");};
}
if (!asmProcess.waitForFinished()) {
    return ...
}
```

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -84,6 +88,8 @@ public slots:
     // Current chosen function symbol
     Data::Symbol m_curSymbol;
+    // Actual application path for current selected symbol
+    QString m_curAppPath;
```
unrelated to this patch, should be part of another one

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -100,6 +106,10 @@ public slots:
+    // Default font size
+    int m_origFontSize;
+    // Version of objdump
+    QString m_objdumpVersion;
```
unused in this patch, either remove or move to the patch where this is being used

milianw commented on src/models/searchdelegate.cpp:
```diff
@@ -0,0 +1,93 @@
+/*
```
looking at what this does, please remove it altogether, this should be handled by using KTextEditor instead

milianw commented on src/models/disassemblymodel.h:
```diff
@@ -0,0 +1,46 @@
+
+#pragma once
+
+#include <QStandardItemModel>
```
sigh... please make it a real model

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -69,7 +69,7 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(FilterAndZoomStack *filterStack,
     ui->searchTextEdit->setPlaceholderText(QLatin1String("Search"));
     ui->asmView->setSelectionMode(QAbstractItemView::SelectionMode::ExtendedSelection);
-
+    ResultsUtil::setupDisassemblyContextMenu(ui->asmView);
```
again, multiple things in one patch

please remove this for now, we should instead use KTextEditor

milianw commented on src/models/CMakeLists.txt:
```diff
@@ -5,6 +5,7 @@ add_library(models STATIC
     callercalleemodel.cpp
     disassemblymodel.cpp
     costdelegate.cpp
+    highlighter.cpp
```
also should be handled by KTextEditor

milianw reviewed (changes requested):
So, please please please try to concentrate on baby steps instead of throwing tons of stuff over the fence that is just not done properly :(

we should really be using KTextEditor here for displaying the data, but quite frankly I would say let's first try to get the basics in and iterate on top of that afterwards then

so please remove everything from this merge request that is not required for the basic disassembly feature

- no mouse wheel zooming
- no custom selection tracking, searching or copy'n'paste features
- no custom highlighting

instead, just get a basic v1 in which displays the costs but make sure at least that part is properly implemented

-> detangle the objdump parsing code from the widget
-> make it run asynchronously instead of synchronously
-> properly show error messages
-> make it possible to configure the objdump binary that should be used, ideally per arch

dknysh commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -1010,11 +1035,17 @@ class PerfParserPrivate : public QObject
 
-        auto frameCallback = [this, &recursionGuard, &sampleCost, type](const Data::Symbol& symbol,
-                                                                        const Data::Location& location) {
+        disassemblyResult.selfCosts.initializeCostsFrom(bottomUpResult.costs);
```
I have entered it due to the difference for LBR frames. If you remember for the moment I have splitted out part related to LBR special processing. But I kept special structures. For other unwining methods you are right there is no difference.

milianw replied:
the LBR fix was at the totally wrong place, it should also apply to the file/line interpretation and not just to the disassembly analysis

regarding the data structures: The `CallerCalleeEntry` should have two maps then, one for the file/line data (i.e. the current `LocationCostMap`), and a second new one for the per-IP data for disassembly. I'd suggest something like this:

```
using SourceLocationCostMap = QHash<QString, LocationCost>;
using OffsetLocationCostMap = QHash<quint64, LocationCost>;
struct CallerCalleeEntry
{
    ...
    SourceLocationCostMap sourceMap;
    OffsetLocationCostMap offsetMap;
};
```

dknysh replied:
Ok, I will put fix to the next iteration of review

dknysh replied:
Fixed, added to the commits

dknysh commented on src/resultsdisassemblypage.cpp:
```diff
@@ -75,6 +72,18 @@ void ResultsDisassemblyPage::clear()
+        ui->asmView->setColumnWidth(event + 1, 100);
+        ui->asmView->header()->setSectionResizeMode(event + 1, QHeaderView::Interactive);
+    }
+    emit model->dataChanged(QModelIndex(), QModelIndex());
```
To repaint model.  I will try to remove

milianw replied:
```
ui->asmView->viewport()->update();
```

but generally, I don't see why this should be required here, it is a code smell that you need to trigger this manually.

dknysh commented:
Milian, 
I have kept only required commits for basic v1 disassembly functionality with costs. 
Fixed your remarks about them.
And implemented point -> properly show error messages

As for the points:
-> detangle the objdump parsing code from the widget
-> make it run asynchronously instead of synchronously
-> make it possible to configure the objdump binary that should be used, ideally per arch

I wrote TODO for them. Probably you could merge v1 without them at least to join with perfparser part.
And then implement them as refactoring.
Or let me know please if you insist on their implementation before merge and which of them are required for this?

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,158 @@
+    ui->splitter->setStretchFactor(1, 1);
+    ui->splitter->setStretchFactor(2, 8);
+
+    m_model = new QStandardItemModel();
```
move up to member init list, pass `this` as parent

milianw replied:
but keep the ui->asmView->setModel(m_model) call here

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,158 @@
+
+void ResultsDisassemblyPage::clear()
+{
+    if (ui->asmView->model() == nullptr)
```
remove, m_model is always valid

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,158 @@
+    if (ui->asmView->model() == nullptr)
+        return;
+
+    int rowCount = ui->asmView->model()->rowCount();
```
operate on m_model also below

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,158 @@
+    // Call objdump with arguments: addresses range and binary file
+    QString processName = m_objdump;
+    QStringList arguments;
+    arguments << QLatin1String("-d") << QLatin1String("--start-address") << QLatin1String("0x") + QString::number(m_curSymbol.relAddr, 16) <<
```
use QStringLiteral here for the string args instead of QLatin1String

but rewrite the `QLatin1String("0x") + QString::number(m_curSymbol.relAddr, 16)` as: `QStringLiteral("0x%1").arg(m_curSymbol.relAddr, 0, 16)`, dito for the relAddr + size arg

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -0,0 +1,158 @@
+            m_model->setItem(row, 0, asmItem);
+            row++;
+        }
+        ui->asmView->setModel(m_model);
```
remove, do it once in the header

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -199,6 +199,11 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
             "objdump");
+
+    if (m_arch.startsWith(QLatin1String("armv8")) || m_arch.startsWith(QLatin1String("aarch64"))) {
+        m_arch = QLatin1String("armv8");
```
here and in the line below: QStringLiteral

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -107,21 +107,66 @@ void ResultsDisassemblyPage::showDisassembly()
+static DisassemblyOutput fromProcess(QString processName, QStringList arguments, QString arch, const Data::Symbol &curSymbol)
+{
+    DisassemblyOutput disassemblyOutput;
+    QByteArray processOutput = QByteArray();
```
remove, unused

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -107,21 +107,66 @@ void ResultsDisassemblyPage::showDisassembly()
+    DisassemblyOutput disassemblyOutput;
+    QByteArray processOutput = QByteArray();
+    if (curSymbol.symbol.isEmpty()) {
+        disassemblyOutput.errorMessage = QLatin1String("Empty symbol ?? is selected");
```
make translatable: `QCoreApplication::tr` and also apply to all the cases below

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -107,21 +107,66 @@ void ResultsDisassemblyPage::showDisassembly()
+    asmProcess.start(processName, arguments);
+
+    bool started = asmProcess.waitForStarted();
+    bool finished = asmProcess.waitForFinished();
```
move down and also inline

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -107,21 +107,66 @@ void ResultsDisassemblyPage::showDisassembly()
+    QProcess asmProcess;
+    asmProcess.start(processName, arguments);
+
+    bool started = asmProcess.waitForStarted();
```
inline into the if (!started)

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -107,21 +107,66 @@ void ResultsDisassemblyPage::showDisassembly()
+    bool started = asmProcess.waitForStarted();
+    bool finished = asmProcess.waitForFinished();
+    if (!started) {
+        if (!arch.startsWith(QLatin1String("arm"))) {
```
move this check up and actually check if the exe can be found by:

```
const auto processPath = QStandpardPaths::findExecutable(processName);
if (processPath.isEmpty()) {
    return //  not found error
}
```

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -107,21 +107,66 @@ void ResultsDisassemblyPage::showDisassembly()
+    disassemblyOutput.output = asmProcess.readAllStandardOutput();
+
+    if (disassemblyOutput.output.isEmpty()) {
+        disassemblyOutput.errorMessage = QLatin1String("Empty output of command ") + processName;
```
tr().arg(processName)

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -107,21 +107,66 @@ void ResultsDisassemblyPage::showDisassembly()
 
-        if (!asmProcess.waitForStarted() || !asmProcess.waitForFinished()) {
+        if (!disassemblyOutput) {
+            KMessageBox::detailedSorry(this, disassemblyOutput.errorMessage, disassemblyOutput.errorMessage);
```
twice the same? that looks wrong? one should be short, one should be long?

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -56,6 +56,87 @@
+            m_model->clear();
+
+            QStringList headerList;
+            headerList.append(QLatin1String("Assembly"));
```
tr

milianw reviewed (changes requested):
thank you we are getting there!

dknysh commented:
Updated accordingly to the remarks.

milianw commented on src/models/data.h:
```diff
@@ -93,6 +93,8 @@ struct DisassemblyResult
     QString extraLibPaths;
     // perf.data path
     QString perfDataPath;
+    // Objdump path
```
remove, this shouldn't be a property of the data

milianw commented on src/mainwindow.cpp:
```diff
@@ -299,7 +305,7 @@ void MainWindow::openFile(const QString& path, bool isReload)
 
     // TODO: support input files of different types via plugins
-    m_parser->startParseFile(path, m_sysroot, m_kallsyms, m_debugPaths, m_extraLibPaths, m_appPath, m_arch);
+    m_parser->startParseFile(path, m_sysroot, m_kallsyms, m_debugPaths, m_extraLibPaths, m_appPath, m_arch, m_objdump);
```
don't pass it to the parser, pass it only to the disassembly page

milianw commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -1268,7 +1268,7 @@ PerfParser::~PerfParser() = default;
 void PerfParser::startParseFile(const QString& path, const QString& sysroot, const QString& kallsyms,
                                 const QString& debugPaths, const QString& extraLibPaths, const QString& appPath,
-                                const QString& arch)
+                                const QString& arch, const QString& objdump)
```
remove

milianw commented on src/parsers/perf/perfparser.cpp:
```diff
@@ -1319,7 +1319,7 @@ void PerfParser::startParseFile(const QString& path, const QString& sysroot, con
     m_callerCalleeResults = {};
     m_events = {};
-    m_disassemblyResult = {path, appPath, extraLibPaths, arch};
+    m_disassemblyResult = {path, appPath, extraLibPaths, arch, objdump};
```
remove

milianw commented on src/parsers/perf/perfparser.h:
```diff
@@ -42,7 +42,7 @@ class PerfParser : public QObject
 
     void startParseFile(const QString& path, const QString& sysroot, const QString& kallsyms, const QString& debugPaths,
-                        const QString& extraLibPaths, const QString& appPath, const QString& arch);
+                        const QString& extraLibPaths, const QString& appPath, const QString& arch, const QString& objdump);
```
remove

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -277,8 +277,12 @@ void ResultsDisassemblyPage::setData(const Data::DisassemblyResult &data)
     m_appPath = data.appPath;
     m_extraLibPaths = data.extraLibPaths;
     m_arch = data.arch.trimmed().toLower();
+    m_objdump = data.objdumpPath;
```
set it on this page from the mainwindow directly, instead of getting it from the data

milianw reviewed (changes requested):
there are still quite a few things that need to be resolved...

dknysh commented on src/resultsdisassemblypage.cpp:
```diff
@@ -107,29 +108,78 @@ void ResultsDisassemblyPage::showDisassembly()
 
-        if (!asmProcess.waitForStarted() || !asmProcess.waitForFinished()) {
+        if (!disassemblyOutput) {
+            KMessageBox::detailedSorry(this, tr("Error. Look at the Details for detailed description."), disassemblyOutput.errorMessage);
```
The length of Sorry text is small. It is the reason I use detailedSorry

dknysh commented:
Updated accordingly your remarks. Please review.

dknysh commented on src/resultsdisassemblypage.cpp:
```diff
@@ -57,6 +57,87 @@
 #include <KColorScheme>
 #include <KMessageBox>
 
+class ObjdumpParser : public QObject
```
Could you clarify how should the test look like and where should be added to? Any examples?

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -57,9 +57,6 @@ ResultsDisassemblyPage::ResultsDisassemblyPage(FilterAndZoomStack *filterStack,
         , ui(new Ui::ResultsDisassemblyPage)
 {
     ui->setupUi(this);
-    ui->splitter->setStretchFactor(0, 1);
```
this is still not answered

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -116,25 +134,73 @@ void ResultsDisassemblyPage::showDisassembly(const QString& processName, const Q
+            const auto addrLine = asmTokens.value(0).trimmed();
+
+            bool ok = false;
+            QString addrLineRef = addrLine;
```
remove

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -116,25 +134,73 @@ void ResultsDisassemblyPage::showDisassembly(const QString& processName, const Q
+
+            bool ok = false;
+            QString addrLineRef = addrLine;
+            const auto addr = QStringRef(&addrLineRef).toULongLong(&ok, 16);
```
remove the QStringRef, operate on addrLine directly

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -116,25 +134,73 @@ void ResultsDisassemblyPage::showDisassembly(const QString& processName, const Q
             QString asmLine = stream.readLine();
             if (asmLine.isEmpty() || asmLine.startsWith(QLatin1String("Disassembly"))) continue;
 
+            QStringList asmTokens = asmLine.split(QLatin1Char(':'));
```
const auto asmTokens = asmLine.splitRef(QLatin1Char(':'));
const auto addrLine = asmTokens[0].trimmed();

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -116,25 +134,73 @@ void ResultsDisassemblyPage::showDisassembly(const QString& processName, const Q
+
+            // Calculate event times and add them in red to corresponding columns of the current disassembly row
+            float costLine = 0;
+            float totalCosts[numTypes];
```
don't use a VLA here, use Data::Cost

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -116,25 +134,73 @@ void ResultsDisassemblyPage::showDisassembly(const QString& processName, const Q
+            float totalCosts[numTypes];
+            auto &entry = m_callerCalleeResults.entry(m_curSymbol);
+
+            QHash<quint64, Data::LocationCost>::iterator i = entry.offsetMap.begin();
```
auto

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -116,25 +134,73 @@ void ResultsDisassemblyPage::showDisassembly(const QString& processName, const Q
+            QHash<quint64, Data::LocationCost>::iterator i = entry.offsetMap.begin();
+            while (i != entry.offsetMap.end()) {
+                quint64 relAddr = i.key();
+                Data::LocationCost locationCost = i.value();
```
totalCosts += i->selfCost;

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -116,25 +134,73 @@ void ResultsDisassemblyPage::showDisassembly(const QString& processName, const Q
+
+            QHash<quint64, Data::LocationCost>::iterator i = entry.offsetMap.begin();
+            while (i != entry.offsetMap.end()) {
+                quint64 relAddr = i.key();
```
unused, remove

at which point this whole manual loop can become:

```
Data::ItemCost totalCosts(0, numTypes);
for (const auto &offset : entry.offsetMap) {
    totalCosts += offset.selfCost;
}
```

but that is by definition equal to the entry cost anyways, so all of this should be replaced by:

```
const auto totalCosts = m_callerCalleeResults.selfCosts.itemCost(entry.id);
```

milianw replied:
and then see below - this should imo be replaced by the actual total cost, not by the function-total-cost to get relative-to-total instead of relative-to-functions numbers

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -116,25 +134,73 @@ void ResultsDisassemblyPage::showDisassembly(const QString& processName, const Q
+
+            auto it = entry.offsetMap.find(addr);
+            if (it != entry.offsetMap.end()) {
+                Data::LocationCost locationCost = it.value();
```
const auto& locationCost = it.value();

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -116,25 +134,73 @@ void ResultsDisassemblyPage::showDisassembly(const QString& processName, const Q
+                    //FIXME QStandardItem stuff should be reimplemented properly
+                    QStandardItem *costItem = new QStandardItem(costInstruction);
+                    costItem->setData(costLine, CostRole);
+                    costItem->setData(totalCosts[event], TotalCostRole);
```
this would now actually compute the total cost relative to the function, which is imo pretty confusing

it would be much better to use the overall total cost instead:

costItem->setData(m_callerCalleeResults->totalCost(event), TotalCostRole);

dknysh replied:
Why do you want relative-to-total instead of relative-to-functions numbers? At least 'perf annotate' computes them relative-to-function. I think we should follow this approach and it should be relative-to-function. If you want relative-to-total values,  probably they can be computed separately and shown nearby column. I can write TODO about them.

milianw replied:
Because the rest of hotspot also presents the data in this way. It would be very confusing if we suddenly show the data differently in this one new view.

and yes, I insist on this change. unification is important

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -36,6 +36,7 @@
 #include <QString>
 #include <QListWidgetItem>
 #include <QProcess>
+#include <QObject>
```
shouldn't be needed, remove

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -100,7 +103,6 @@ void ResultsDisassemblyPage::showDisassembly()
     // Show empty tab when selected symbol is not valid
     if (m_curSymbol.symbol.isEmpty()) {
         clear();
-        return;
```
why is this removed?

dknysh replied:
To show diagnostics for Empty symbol.

milianw commented on src/resultsdisassemblypage.h:
```diff
@@ -56,6 +56,13 @@ struct DisassemblyOutput
+        quint64 addr = 0;
+        QString disassembly;
+    };
+    QVector<DisassemblyLine> disassemblyLines;
```
use this to replace the `QByteArray output`

dknysh replied:
What do you mean? I need them both - QByteArray output and QVector<DisassemblyLine> disassemblyLines.

dknysh commented on src/resultsdisassemblypage.h:
```diff
@@ -100,3 +108,5 @@ class ResultsDisassemblyPage : public QWidget
     CostDelegate *m_costDelegate;
 };
+
+Q_DECLARE_TYPEINFO(DisassemblyOutput::DisassemblyLine, Q_MOVABLE_TYPE);
```
Could you please repeat 'still open' remark about output? I can not find initial remark.

dknysh commented:
Updated accordingly your remarks.

dknysh commented:
Could you please review?

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -116,25 +134,61 @@ void ResultsDisassemblyPage::showDisassembly(const QString& processName, const Q
+            const auto addrLine = asmTokens[0].trimmed();
+
+            bool ok = false;
+            const auto addr = QStringRef(addrLine).toULongLong(&ok, 16);
```
the explicit QStringRef here isn't needed, addrLine is a QStringRef already

dknysh replied:
I don't see initial question about addTab but as I remember it was your concern about output uniqueness. I confirm that the output is proper and unique for each onJumpToDisassembly and addTab. And before any symbol selected I prefer not to show the tab.

milianw commented on src/resultsdisassemblypage.cpp:
```diff
@@ -245,3 +235,30 @@ void ResultsDisassemblyPage::setCostsMap(const Data::CallerCalleeResults& caller
+
+void ResultsDisassemblyPage::objdumpParse(DisassemblyOutput& disassemblyOutput)
+{
+    QVector<DisassemblyOutput::DisassemblyLine> disassemblyLines;
```
this is the wrong place for this code, this parsing should be done within the DisassemblyOutput

i.e. you parse the raw `QByteArray output` and only return the parsed `QVector<DisassemblyOutput::DisassemblyLine>` then. The raw output isn't needed afterwards anymore

milianw replied:
i.e. this should happen within `DisassemblyOutput fromProcess(...)` already

dknysh commented:
Updated accordingly your remarks. Could you please review?

milianw commented:
thank you @dknysh, this looks pretty good now, besides a few minor nits that I'll take care of myself. I'll try to get this in today, together with the perfparser changes

milianw commented:
Oh, I notice that you didn't rebase in a while apparently, can you please do that and fix the merge conflicts accordingly?

dknysh commented:
I can not build last KDAB due KF5 dependency. What should I install on Ubuntu 19 to resolve?

dknysh commented:
I can not build latest KDAB due dependency on KF5 notifications.

The following packages have unmet dependencies:
 libkf5notifications-dev : Depends: libkf5notifications5 (= 5.62.0-0ubuntu1) but it is not going to be installed
                           Recommends: libkf5notifications-doc (= 5.62.0-0ubuntu1) but 5.65.0-0ubuntu1~ubuntu19.10~ppa1 is to be installed
E: Unable to correct problems, you have held broken packages.

How can I install it on Ubuntu 19?

dknysh commented:
Ok, I will rebase.

dknysh commented:
I have rebased. Could you please merge?

milianw commented:
I've now merged this and started cleaning up the code - please, the next time at least try not to add new compile warnings

dknysh commented:
Yes, sure, thanks.
